include/kernels/gauge_wilson_flow.cuh

```diff
@@ -0,0 +1,192 @@
+#include <gauge_field_order.h>
+#include <index_helper.cuh>
+#include <quda_matrix.h>
+#include <kernels/gauge_utils.cuh>
+
+namespace quda
+{
+
+  template <typename Float_, int nColor_, QudaReconstructType recon_, int wFlowDim_> struct GaugeWFlowArg {
+    using Float = Float_;
+    static constexpr int nColor = nColor_;
+    static_assert(nColor == 3, "Only nColor=3 enabled at this time");
+    static constexpr QudaReconstructType recon = recon_;
+    static constexpr int wFlowDim = wFlowDim_;
+    typedef typename gauge_mapper<Float,recon>::type Gauge;
+
+    Gauge out;
+    Gauge temp;
+    Gauge in;
+
+    int threads; // number of active threads required
+    int X[4];    // grid dimensions
+    int border[4];
+    const Float epsilon;    
+    
+    GaugeWFlowArg(GaugeField &out, GaugeField &temp, GaugeField &in, Float epsilon) :
+      out(out),
+      in(in),
+      temp(temp),
+      threads(1),
+      epsilon(epsilon)
+    {
+      for (int dir = 0; dir < 4; ++dir) {
+        border[dir] = in.R()[dir];
+        X[dir] = in.X()[dir] - border[dir] * 2;
+        threads *= X[dir];
+      }
+      threads /= 2;
+    }
+  };
+
+  // Wilson Flow as defined in https://arxiv.org/abs/1006.4518v3 
```

Comment:
The three kernels here, while all different, clearly have much boiler plate in comment, e.g., all the indexing.  How about instead a single templated kernel, with a switch statement inside doing either `W1`, `W2` or `Wt` paths?  Keep separate tune keys for the different kernel types of course.

Reply:
Just looking for clarification because my C++ isn't that great. There will be a single `Tunable` instance, such as:
```C++
template <typename Float, int nColor, QudaReconstructType recon> class GaugeWFlow : TunableVectorYZ
```
which has members:
```C++
GaugeWFlowW1(GaugeField &out, GaugeField &temp, GaugeField &in, double epsilon);
GaugeWFlowW2(GaugeField &out, GaugeField &temp, GaugeField &in, double epsilon);
GaugeWFlowVt(GaugeField &out, GaugeField &temp, GaugeField &in, double epsilon);
```
so that the tuning strings are separated, and a case/swicth on
```C++
computeWFlowStep[W1,W2,Vt]<<<tp.grid, tp.block, tp.shared_bytes>>>(arg);
```
Is that the gist of it?

Reply:
I'm thinking something like coarse_op_preconditioned.cu:
* Although there's only one kernel, there is a templated option which controls whether the kernel does a norm or the actual calculation.  I'm thinking here we simply have a template enum parameter.  Here's a cartoon.
```c++
template <FlowType type_>
struct Arg {
  static constexpr FlowType type = type_;
};

template <Arg arg>
__global__ void foo(Arg arg)
{
  // first do common indexing etc.

  // now specialize according to type...
  switch(arg.type) {
  case W1type: // do W1 compute
  case W2type: // do W2 compute
  case WTtype: // do WT compute
  }
}
```
* Same `Tunable` class is used for both, and there's an external knob to control which compute-type is being done.  And there is an explicit string added to the `TuneKey` so to break the degeneracy of the tuning.

Reply:
OK. I didn’t realise that one could update the halo regions from inside a kernel call. That gets rid of the calls to the gauge exchange in the interface. With that it’s simple to put the nStep loop in the kernel, exchange ghost zones if using MPI, and templetise a WFlowStep call on step type. I’ll update the smearing routines to follow this style in a separate pull. Then I’ll push my updates to over-relaxation gauge fixing.